src/react/utils/useHydrateAtoms.ts

```diff
@@ -4,11 +4,20 @@ import type { WritableAtom } from '../../vanilla.ts'
 type Store = ReturnType<typeof useStore>
 type Options = Parameters<typeof useStore>[0]
 type AnyWritableAtom = WritableAtom<unknown, any[], any>
+type AtomTuple<A = AnyWritableAtom, V = unknown> = readonly [A, V]
+
+type InferAtoms<T extends Iterable<AtomTuple>> = {
+  [K in keyof T]: T[K] extends AtomTuple<infer A>
+    ? A extends AnyWritableAtom
+      ? AtomTuple<A, ReturnType<A['read']>>
+      : T[K]
+    : never
+}
 
 const hydratedMap: WeakMap<Store, WeakSet<AnyWritableAtom>> = new WeakMap()
 
-export function useHydrateAtoms(
-  values: Iterable<readonly [AnyWritableAtom, unknown]>,
+export function useHydrateAtoms<T extends Iterable<AtomTuple>>(
+  values: InferAtoms<T>,
```

Comment:
Are there any other tests you'd like me to add for this change?

Reply:
Added this test after tweaking type to use `any` instead of `never` (needed to continue to support `Map`):
https://github.com/pmndrs/jotai/blob/b9ba05b704ab29cdfc9ba30133668e69cb9b97b4/tests/react/utils/useHydrateAtoms.test.tsx#L8

Reply:
I think we should add tests for type errors, with expectType or ts-expect-error.

Reply:
I added one test using `ts-expect-error` in commit https://github.com/pmndrs/jotai/pull/1898/commits/6b03f946edb41d4dcb70bc9686886820c4390e6c